From f0bb7a2d56d5bbabd1ad838b7fb53aed74e26807 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:35:52 -0500
Subject: [fs] overlayfs: embed middle into overlay_readdir_data

Message-id: <20141125163552.18115.87095.stgit@warthog.procyon.org.uk>
Patchwork-id: 100939
O-Subject: [PATCH 39/47] overlayfs: embed middle into overlay_readdir_data
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Al Viro <viro@zeniv.linux.org.uk>

same story...

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Original-commit: db6ec212b53abc29a5bb6ac8c810010fc28d5191
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/overlayfs/readdir.c b/fs/overlayfs/readdir.c
index 453c4bf..7ec7f2a 100644
--- a/fs/overlayfs/readdir.c
+++ b/fs/overlayfs/readdir.c
@@ -43,7 +43,7 @@ struct ovl_readdir_data {
  bool is_merge;
  struct rb_root root;
  struct list_head *list;
- struct list_head *middle;
+ struct list_head middle;
  int count;
  int err;
 };
@@ -144,13 +144,13 @@ static int ovl_fill_lower(struct ovl_readdir_data *rdd,
 
  p = ovl_cache_entry_find(&rdd->root, name, namelen);
  if (p) {
-  list_move_tail(&p->l_node, rdd->middle);
+  list_move_tail(&p->l_node, &rdd->middle);
  } else {
   p = ovl_cache_entry_new(name, namelen, ino, d_type);
   if (p == NULL)
    rdd->err = -ENOMEM;
   else
-   list_add_tail(&p->l_node, rdd->middle);
+   list_add_tail(&p->l_node, &rdd->middle);
  }
 
  return rdd->err;
@@ -284,7 +284,6 @@ static inline int ovl_dir_read_merged(struct path *upperpath,
           struct list_head *list)
 {
  int err;
- struct list_head middle;
  struct ovl_readdir_data rdd = {
   .ctx.actor = ovl_fill_merge,
   .list = list,
@@ -308,11 +307,10 @@ static inline int ovl_dir_read_merged(struct path *upperpath,
    * Insert lowerpath entries before upperpath ones, this allows
    * offsets to be reasonably constant
    */
-  list_add(&middle, rdd.list);
-  rdd.middle = &middle;
+  list_add(&rdd.middle, rdd.list);
   rdd.is_merge = true;
   err = ovl_dir_read(lowerpath, &rdd);
-  list_del(&middle);
+  list_del(&rdd.middle);
  }
 out:
  return err;
-- 
1.7.1